### PR TITLE
Add MagNet based detector

### PR DESCRIPTION
**Pull request by VRehnberg**
Started on adaption of [MagNet](https://www.semanticscholar.org/paper/MagNet%3A-A-Two-Pronged-Defense-against-Adversarial-Meng-Chen/63a010c69f00e65c946a68b546bbd42cbed03564) detector.

Original implementation is basically:
1. Train autoencoder on clean input data
2. Anomaly if
    1. Reconstruction loss is high or
    2. KL divergence between output logits for input and output logits for reconstructed input (I call this "stability loss")

I've adapted this so that autoencoders are trained on the activations instead of the input (similar to AbstractionDetector). For linear layers the current autoencoders is simply linear layers while for conv layers I used something similar to the one used in the paper but other activation functions. This is probably not reasonable.

Todo:
 - [x] Implement reconstruction loss for activation layers
 - [ ] Implement stability loss (i.e. will probably need to intervene on activations and have a forward pass per used activation layer)
 - [x] Make it run with any AutoEncoder.
 - [ ] Find reasonable auto encoders to use.

Depending on
- [x] https://github.com/ejnnr/cupbearer/pull/14

### CODE REVIEW

**Comment by ejnnr** [comment]
I've changed the base branch to `pytorch` for now to make it easier for me to understand the diff (we can change it back when #14 gets merged). I also took a brief look at the code but not too detailed yet, let me know once an actual review would be useful. Seems like there might be ways to share some code with abstractions, but might be easiest to figure out the best design here a bit later, duplicating some code is fine by me for now.

I like autoencoders on activations, it's a good instance of the general "OOD detection on activations" approach! (wouldn't call it MagNet, but looks like your code doesn't either)

I actually think the stability loss from the original paper is also an interesting method. It doesn't directly use model internals, but it has some similar effects, e.g. it could detect backdoors/... that differ only a tiny but crucial amount from clean inputs. It probably fails for triggers that are very big/robust, but from an AI safety perspective those feel less scary and hopefully can be defeated with other methods. Overall I'd guess this method isn't the end game, but it's a good way of checking that supposedly better methods actually do anything difficult.

A stability loss on activations is also interesting, I think depending on the implementation details, this would be very close to a form abstractions. (Though not quite the version I've implemented.) Because of that, and because it seems like it might be more work to implement, I think it would be fine to skip that one for now. (Or if you do want it right away, maybe we really should think about sharing more code between abstractions and autoencoders)

**Comment by VRehnberg** [comment]
I'd say it is good enough now. The rest can be added when needed. Two things worth noting:
 - I've checked that the code runs, not if it is in any way useful.
 - I don't really like the names of stuff now.
      - AutoencoderAbstraction: questionable whether this should be called an abstraction or not
      - AcyclicAbstraction: alt. PredictingAbstraction, DAGAbstraction, just Abstraction, ...
      - Abstraction: what else could be common between these methods? LearnedCommutativeProperty, ActivationMap, ... 

I'll ask for a review when pytorch branch has been merged.

**Comment by VRehnberg** [comment]
Also test_abstractions has a lot of code duplication as is, should perhaps do something about that.

**Review by ejnnr — COMMENTED** [review]
This looks good! I agree the names could use some work and that re-using the defaults from `HookedModel` would be nice. I think we might not always want linear maps as tau maps though, e.g. sometimes convolutions. Also for the older abstractions, we need abstract computational steps, which seem trickier to infer automatically. If you don't see a good way to solve this, I think keeping some duplication isn't the worst thing ever, it feels like we should fix this once we implement yet another similar detector and start seeing patterns more clearly. It might be good to have a check in the default abstraction functions though, whether the names are the same as `default_names`. (Just so that if we every change default names, we remember to change default abstractions too)

File: src/cupbearer/detectors/abstraction/__init__.py [hunk 1]
```diff
@@ -37,25 +38,42 @@ def setup_and_validate(self):
...
+    def build(self, model: HookedModel) -> AutoencoderAbstraction:
+        return AutoencoderAbstraction.get_default(
             model,
             self.size_reduction,
         )
 
 
 ABSTRACTIONS = {
-    "simple": AbstractionConfig,
+    "simple": AcyclicAbstractionConfig,
```
[the remark this answers is not in the document]
Comment (VRehnberg) [reply src/cupbearer/detectors/abstraction/__init__.py]: Yes, I like that name much more.

File: src/cupbearer/detectors/abstraction/abstraction_detector.py [hunk 2]
```diff
@@ -6,47 +7,86 @@
...
+    AutoencoderAbstraction,
+)
 from cupbearer.detectors.anomaly_detector import (
     ActivationBasedDetector,
 )
 from cupbearer.models import HookedModel
 from cupbearer.utils.optimizers import OptimizerConfig
 
 
+def per_layer(func: Callable):
```
[the remark this answers is not in the document]
Comment (VRehnberg) [reply src/cupbearer/detectors/abstraction/abstraction_detector.py:22]: Thanks

**Comment by VRehnberg** [comment]
> This looks good! I agree the names could use some work and that re-using the defaults from `HookedModel` would be nice. I think we might not always want linear maps as tau maps though, e.g. sometimes convolutions. Also for the older abstractions, we need abstract computational steps, which seem trickier to infer automatically. If you don't see a good way to solve this, I think keeping some duplication isn't the worst thing ever, it feels like we should fix this once we implement yet another similar detector and start seeing patterns more clearly. It might be good to have a check in the default abstraction functions though, whether the names are the same as `default_names`. (Just so that if we every change default names, we remember to change default abstractions too)

I made some changes to make this easier (and I think could be useful for forward hooks). Cleaning up default for MLP would be quite easy now, but already for CNN the logic becomes more confusing than before to allow for arbitrary name graphs. Probably it would be easiest to use something like [torchinfo](https://github.com/tyleryep/torchinfo) to get output sizes of layers.

**Comment by VRehnberg** [comment]
@ejnnr Are there any other changes you would like to see?

**Comment by ejnnr** [comment]
> Are there any other changes you would like to see?

No, looks good, merging this!
